### PR TITLE
Make tidy check that opening and closing braces that begin a line do …

### LINKED ISSUE #16486: tidy: Make tidy check that opening and closing braces that begin a line do so with proper alignment.
I think it makes sense if we enforce the invariant that at least a `{` or `}` that begins a line should do so with a multiple of 4 spaces to the left. That'd prevent stuff like https://github.com/servo/servo/pull/16293#discussion_r111661247 from landing.

@Wafflespeanut wanna mentor this one? or should I? (should be easy)

### PR DESCRIPTION
**Pull request by coalman**
…so with proper alignment.

---

- [X] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [X] These changes fix #16486 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16495)

### CODE REVIEW

**Comment by jdm** [comment]
`./mach test-tidy --all` reports a bunch of preexisting issues that will need to be corrected:
```
./components/devtools/actors/network_event.rs:368: space before } is not a multiple of 4

./components/devtools/actors/network_event.rs:416: space before } is not a multiple of 4

./components/layout/block.rs:1698: space before } is not a multiple of 4

./components/layout/block.rs:1700: space before } is not a multiple of 4

./components/layout/display_list_builder.rs:236: space before } is not a multiple of 4

./components/layout/table_wrapper.rs:340: space before } is not a multiple of 4

./components/layout/table_wrapper.rs:342: space before } is not a multiple of 4

./components/net/hsts.rs:117: space before } is not a multiple of 4

./components/net/mime_classifier.rs:168: space before } is not a multiple of 4

./components/net/mime_classifier.rs:818: space before } is not a multiple of 4

./components/profile/mem.rs:79: space before } is not a multiple of 4

./components/profile/time.rs:288: space before } is not a multiple of 4

./components/script/script_thread.rs:1969: space before } is not a multiple of 4

./components/script/webdriver_handlers.rs:60: space before } is not a multiple of 4

./components/script/dom/blob.rs:361: space before } is not a multiple of 4

./components/script/dom/document.rs:1337: space before } is not a multiple of 4

./components/script/dom/document.rs:2363: space before } is not a multiple of 4

./components/script/dom/dommatrixreadonly.rs:40: space before } is not a multiple of 4

./components/script/dom/element.rs:1083: space before } is not a multiple of 4

./components/script/dom/element.rs:2602: space before } is not a multiple of 4

./components/script/dom/formdata.rs:106: space before } is not a multiple of 4

./components/script/dom/htmlareaelement.rs:130: space before } is not a multiple of 4

./components/script/dom/htmlareaelement.rs:150: space before } is not a multiple of 4

./components/script/dom/htmlimageelement.rs:663: space before } is not a multiple of 4

./components/script/dom/htmlimageelement.rs:665: space before } is not a multiple of 4

./components/script/dom/htmlimageelement.rs:670: space before } is not a multiple of 4

./components/script/dom/htmlimageelement.rs:672: space before } is not a multiple of 4

./components/script/dom/htmlimageelement.rs:685: space before } is not a multiple of 4

./components/script/dom/htmlimageelement.rs:687: space before } is not a multiple of 4

./components/script/dom/htmlimageelement.rs:691: space before } is not a multiple of 4

./components/script/dom/htmlimageelement.rs:692: space before } is not a multiple of 4

./components/script/dom/htmlimageelement.rs:693: space before } is not a multiple of 4

./components/script/dom/htmlselectelement.rs:181: space before } is not a multiple of 4

./components/script/dom/htmlselectelement.rs:183: space before } is not a multiple of 4

./components/script/dom/htmlselectelement.rs:184: space before } is not a multiple of 4

./components/script/dom/htmlselectelement.rs:186: space before } is not a multiple of 4

./components/script/dom/htmlselectelement.rs:187: space before } is not a multiple of 4

./components/script/dom/node.rs:2386: space before } is not a multiple of 4

./components/script/dom/storage.rs:95: space before } is not a multiple of 4

./components/script/dom/treewalker.rs:460: space before } is not a multiple of 4

./components/script/dom/webglrenderingcontext.rs:1798: space before } is not a multiple of 4

./components/script/dom/webglrenderingcontext.rs:1825: space before } is not a multiple of 4

./components/script/dom/bindings/trace.rs:731: space before } is not a multiple of 4

./components/script_layout_interface/reporter.rs:48: space before } is not a multiple of 4

./components/selectors/matching.rs:130: space before } is not a multiple of 4

./components/selectors/matching.rs:131: space before } is not a multiple of 4

./components/selectors/parser.rs:377: space before } is not a multiple of 4

./components/selectors/parser.rs:378: space before } is not a multiple of 4

./components/selectors/parser.rs:379: space before } is not a multiple of 4

./components/selectors/parser.rs:384: space before } is not a multiple of 4

./components/selectors/parser.rs:388: space before } is not a multiple of 4

./components/selectors/parser.rs:430: space before } is not a multiple of 4

./components/style/animation.rs:384: space before } is not a multiple of 4

./components/style/attr.rs:165: space before } is not a multiple of 4

./components/style/values/specified/basic_shape.rs:404: space before } is not a multiple of 4

./components/style/values/specified/basic_shape.rs:599: space before } is not a multiple of 4

./components/style/values/specified/basic_shape.rs:611: space before } is not a multiple of 4

./components/style/values/specified/length.rs:1025: space before } is not a multiple of 4

./components/style/values/specified/mod.rs:557: space before } is not a multiple of 4

./components/webdriver_server/lib.rs:98: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2029: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2031: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2037: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2052: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2053: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2060: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2061: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2070: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2087: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2088: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2089: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2100: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2101: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2102: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2103: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2104: space before } is not a multiple of 4

./ports/geckolib/glue.rs:2105: space before } is not a multiple of 4

./ports/glutin/window.rs:961: space before } is not a multiple of 4

./tests/unit/style/media_queries.rs:291: space before } is not a multiple of 4

./tests/unit/style/viewport.rs:318: space before } is not a multiple of 4

./tests/unit/style/viewport.rs:330: space before } is not a multiple of 4

./tests/unit/style/viewport.rs:345: space before } is not a multiple of 4

./tests/unit/style/viewport.rs:359: space before } is not a multiple of 4

./tests/unit/style/properties/serialization.rs:236: space before } is not a multiple of 4

./tests/unit/style/properties/serialization.rs:269: space before } is not a multiple of 4
```

**Comment by emilio** [comment]
Hi, thanks for looking into this @coalman!

Looking at the existing failures, I didn't think of the use case of passing a struct or closure as an argument:

```rust
example.bar()
       .baz(|| {
            // ...
       });
```

```rust
function(a,
         B {
             member: 2,
         });
```

We could force 4-spaces for the first case (moving `.baz(|| ...` also to the left), and do something similar for function arguments, but those changes feel a bit more opinionated, so perhaps we should allow the case of being four spaces to the left of the previous line as an special case.

What do you think?

**Comment by emilio** [comment]
I believe that should account for all the false positives, the rest of them should be fixed for this to land.

Thanks again!

**Comment by jdm** [comment]
Additionally, we should add automated tests for this. See `python/tidy/servo_tidy_tests/test_tidy.py`, and run them with `./mach test-tidy --self-test`.

**Comment by coalman** [comment]
@emilio I added the changes you suggested and `./mach test-tidy --all` doesn't show any false positives anymore. I added an edge case to the tidy self-test as well.

File: python/tidy/servo_tidy/tidy.py [hunk 1]
```diff
@@ -450,6 +450,8 @@ def check_rust(file_name, lines):
     prev_crate = {}
     prev_mod = {}
     prev_feature_name = ""
+    indent = 0
+    prev_indent = 0
```
Comment (coalman) [inline python/tidy/servo_tidy/tidy.py]: Having `prev_indent`, `indent`, and `current_indent` is kind of irking me. I'm definitely open to suggestions.

**Comment by emilio** [comment]
@bors-servo r+

Looks great, thanks!

**Comment by Wafflespeanut** [comment]
@bors-servo r-

First two commits need squash.

**Comment by coalman** [comment]
Sorry about that, squashed the first two commits.

**Comment by Wafflespeanut** [comment]
Thank you!

@bors-servo r=emilio

**Pull request merged**